plugins/changevein.cpp

```diff
@@ -24,6 +24,188 @@ DFHACK_PLUGIN("changevein");
+constexpr uint8_t WEST = 3;
+constexpr uint8_t NORTHEAST = 0;
+constexpr uint8_t SOUTHEAST = 1;
+constexpr uint8_t SOUTHWEST = 2;
+constexpr uint8_t NORTHWEST = 3;
+
+static const df::coord2d DIRECTION_OFFSETS[] = { df::coord2d(0, -16), df::coord2d(16, 0), df::coord2d(0, 16), df::coord2d(-16, 0)  };
+static const df::coord2d DIAGONAL_OFFSETS[] = { df::coord2d(16, -16), df::coord2d(16, 16), df::coord2d(-16, -16), df::coord2d(-16, 16) };
+
+struct VeinEdgeBitmask
+{
+    int32_t mat;
+    // North: 0 | East: 1 | South: 2 | West: 3
+    uint16_t edges[4];
+    // NE: 0 | SE: 1 | SW: 2 | NW: 3
+    bool corners[4];
+
+    VeinEdgeBitmask(df::block_square_event_mineralst* event)
+    {
+        mat = event->inorganic_mat;
+
+        // North & South
+        edges[NORTH] = event->tile_bitmask.bits[0];
+        edges[SOUTH] = event->tile_bitmask.bits[15];
+
+        // East & West
+        edges[EAST] = 0;
+        edges[WEST] = 0;
+        for (uint8_t i = 0; i < 16; i++)
+        {
+            // East-most tile is represented by the largest bit
+            edges[EAST] |= (event->tile_bitmask.bits[i] & 0b1000000000000000) >> (15 - i);
+            // West-most tile is represented by the smallest bit
+            edges[WEST] |= (event->tile_bitmask.bits[i] & 0b0000000000000001) << i;
+        }
+
+        corners[NORTHEAST] = edges[NORTH] & 0b1000000000000000;
+        corners[SOUTHEAST] = edges[SOUTH] & 0b1000000000000000;
+        corners[SOUTHWEST] = edges[SOUTH] & 0b0000000000000001;
+        corners[NORTHWEST] = edges[NORTH] & 0b0000000000000001;
+    }
+
+    void SmudgeBitmask()
+    {
+        for (uint8_t i = 0; i < 4; i++)
+        {
+            edges[i] |= (edges[i] << 1) | (edges[i] >> 1);
+        }
+    }
+
+    bool Merge(VeinEdgeBitmask& other)
+    {
+        if (other.mat != this->mat)
+            return false;
+
+        for (uint8_t i = 0; i < 4; i++)
+        {
+            this->edges[i] |= other.edges[i];
+            this->corners[i] |= other.corners[i];
+        }
+
+        return true;
+    }
+
+    bool Merge(df::block_square_event_mineralst* event)
+    {
+        VeinEdgeBitmask temp = VeinEdgeBitmask(event);
+        return this->Merge(temp);
+    }
+};
+
+static void ChangeSameBlockVeins(df::map_block* block, df::block_square_event_mineralst* event, VeinEdgeBitmask& mask, int32_t matIndex)
+{
+    for (size_t i = 0; i < block->block_events.size(); i++)
+    {
+        df::block_square_event* evt = block->block_events[i];
+        if (evt == event)
+            continue;
+        if (evt->getType() != block_square_event_type::mineral)
+            continue;
+
+        df::block_square_event_mineralst* cur = (df::block_square_event_mineralst*)evt;
+        if (cur->inorganic_mat == mask.mat)
+        {
+            for (uint8_t j = 0; j < 15; j++)
+            {
+                uint16_t bitmask = cur->tile_bitmask.bits[j] | cur->tile_bitmask.bits[std::max(j - 1, 0)] | cur->tile_bitmask.bits[std::min(j + 1, 15)];
```

Comment:
isn't j-1 == 255 since it's a uint8_t and has a starting value of 0?

Reply:
Great catch! Got used to using uint as if it was an int and didn't notice.